### ISSUE TITLE
Customize editor while drawing

### ISSUE THREAD

**Issue by ghost** [annotorious/annotorious#431]
Hi @rsimon 

I was playing around with the Annotorious V3 (for OSD) and had a few questions. While there are options to change fill and stroke colors for the finished shape in  OpenSeadragonAnnotator which is great, I wanted to ask if I could customize those properties while I'm drawing (i.e. dragging with mouse). Another question I had was, once the shape is drawn it goes into "editor" mode for that shape. I'd have to click outside the shape once to go back to drawing again. Is it possible to change this behavior and draw the next shape without having to click outside? Thanks!

**Comment by rsimon** [comment]
Hi, I think you should be able to programmatically clear the selection in the `createAnnotation` handler. This way the annotation would instantly deselect, and you could start drawing again.

About the shape styles: for now, you couls change the initial style by overriding the CSS. I think I'd like to add config options for this eventually, to make this easier or at least more obvious. Let me know if this helps, or if you have any suggestions on what kind of API you think would make sense.

**Comment by ghost** [comment]
The clear selection works, thanks! It would be great if we could change the current drawing tool once the annotation is created (like deselecting rect tool back to navigation). 

I think adding config options sounds like a better alternative (maybe alongside the existing "stroke" and "fill") too. For now, is it possible to pass style props (like fill, stroke) to the scripts that draw the shapes (I think the rubberband code?)

**Comment by rsimon** [comment]
The fill/stroke-etc. settings are supposed to apply to the rubberband drawing tools as well. However I think that may not yet be fully implemented. (It definitely isn't in the 'image' version of Annotorious. But might be in the OpenSeadragon version?)

About the behavior: yes, people want different behaviors. Some prefer to have the tool stay enabled after drawing is complete. Others prefer the tool to snap back to "navigate" mode. (In fact I've implemented both options for different clients.) In Annotorious, behavior is going to stay as it is: the tool will not change automatically after drawing is complete. But, likewise, that's easy to control yourself, by simply changing the drawing mode/tool inside the `createAnnotation` or `updateAnnotation` handler.

**Comment by ghost** [comment]
> The fill/stroke-etc. settings are supposed to apply to the rubberband drawing tools as well. However I think that may not yet be fully implemented. (It definitely isn't in the 'image' version of Annotorious. But might be in the OpenSeadragon version?)
 
I'm only able to find the computeStyle variable passed as style in shapes and editors folders but not in tools. I wonder if there's a way to modify the tools scripts to get the same styles? 

> About the behavior: yes, people want different behaviors. Some prefer to have the tool stay enabled after drawing is complete. Others prefer the tool to snap back to "navigate" mode. (In fact I've implemented both options for different clients.) In Annotorious, behavior is going to stay as it is: the tool will not change automatically after drawing is complete. But, likewise, that's easy to control yourself, by simply changing the drawing mode/tool inside the `createAnnotation` or `updateAnnotation` handler.

Oh yeah, I was thinking the same. Thanks!

**Comment by rsimon** [comment]
Ah - well in that case, the style isn't applied in either the OSD version nor the image version yet :-) Yes, I'm definitely going to apply the styles to the tools as well, by the time of the official v3 release.

**Comment by ghost** [comment]
Sounds good! Just for testing, could you maybe point me in the direction of how I could do it myself if it's not too complicated?  Also, if you don't mind me asking, how can I implement an "eraser" tool? As in, when I mousedown and drag, the shapes over which it passes would be removed. I see that there are clearAnnotations() and removeAnnotation(), but I'm not exactly sure how I could use them.

**Comment by rsimon** [comment]
It would actually be quite straightforward. You'd just have to repeat the same steps for the Tools, as they are already implemented for the Editors. (Compute the style and apply it in the Svelte markup templates.)

Re eraser tool: that would be more involved. You would have to build your own logic to handle the mouse events. And you'd have to work with the `anno.state.store` object, which would allow you to query the annotations for a particular X/Y coordinate. You could then delete the annotation simply using  `anno.state.store.deleteAnnotation(id)`.

**Comment by ghost** [comment]
> It would actually be quite straightforward. You'd just have to repeat the same steps for the Tools, as they are already implemented for the Editors. (Compute the style and apply it in the Svelte markup templates.)

Thanks, it works!  I did observe an interesting behavior. When I change the fill and stroke in the props, their values for all the existing annotations are also changed. Looks like they are overriden by the style props. I could be doing something wrong here and probably have to use setStyle() each time I wish to change the next shape props?  

> Re eraser tool: that would be more involved. You would have to build your own logic to handle the mouse events. And you'd have to work with the `anno.state.store` object, which would allow you to query the annotations for a particular X/Y coordinate. You could then delete the annotation simply using `anno.state.store.deleteAnnotation(id)`.

Indeed sounds involved. I tried removing them by clicking on them (using selection). But for dragging, I imagine that we'd have to iterate over annotations, get their bounds and constantly check for intersections? There's also preventing navigation when trying to drag.

**Comment by rsimon** [comment]
I might be misunderstanding your use case. But assuming that you are using React, then, yes, the style will apply to all annotations, and update for all as soon as you change the prop. This is why you can either apply a style object (valid for all), or a _function_ that take an annotation as input and returns a style based on your own logic for assigning styles to annotations.

> But for dragging, I imagine that we'd have to iterate over annotations, get their bounds and constantly check for intersections?

The `state.store` has a `getAt` method that you could use. But that, I'm afraid, isn't documented yet. Until I get around to writing proper documentation, you'd [need to read into the source code](https://github.com/annotorious/annotorious/blob/main/packages/annotorious/src/state/ImageAnnotatorState.ts#L55-L58) directly.

**Comment by ghost** [comment]
Thanks! My bad, I wasn't clear with my use case. I was thinking of having a color input that when changed would apply to the subsequent shapes you would draw, without affecting the ones already present. Also, it would be nice to change the color of the selected annotation separately. I couldn't find a method that selectively changes styles though. In React, setStyle() function likely applies styles to all existing shapes. 

The getAt() seems to take (x,y) coordinates, but I'm unsure what to put in them. In the OSD docs, there's an [example](https://openseadragon.github.io/examples/viewport-coordinates/) that shows three types, but I don't know if any of those should be used. Right now, I'm using useSelection() hook to get the selected annotation and remove it.

**Comment by rsimon** [comment]
Ok, that's what I thought. I know it's a tricky initially. But your use case is exactly what the function-type styles are for. You need to store the color of the annotation somewhere (usually an annotation body). But you might just as well keep a table of IDs -> colors outside. The important part is that the style function is what turns whatever condition you apply to a style.

I wrote tutorial a while back here: https://observablehq.com/@rsimon/writing-annotorious-formatters-pt-1

That's for Annotorious v2, and what's now simply the "style function" used to be called a formatter. It's not exactly the same thing, but the same principle. Hope this helps!

Re `getAt(x,y)`. Yes good question. The x/y coordinates are OpenSeadragon's image coordinates. (Actual pixels of your image in original full resolution.)

**Comment by ghost** [comment]
So, I'd probably store the styles for each annotation in the "bodies" prop and might have to access that prop inside the shapes and editors folders to override the styles? And the "style function" would help with setting new styles in the bodies prop? 

On a separate note, I had issues with the usage of anno.on('updateAnnotation',updateAnnotation). The callback doesn't run when I move/resize the shapes. Could you explain what this event does? Right now, I'm able to store the created and deleted shapes (with the help of create, remove events) in a separate array, but cannot access and store the updated shapes. 

Also, I noticed the component re-renders many times due to moving/resizing shapes which I think hurts performance in React. Is there a way to mitigate this issue? Thanks!

**Comment by rsimon** [comment]
Yes, you'd store information about the style in a body. (Could just be a tag. E.g. "blue" vs. "red".) Then in the style function you'd check the body for the tag value, and return the appropriate style (strike, fill, etc.) object.

**Comment by rsimon** [comment]
The update event fires after you make a change, and deselect the annotation. (Or after a period of idleness, if the autoSave init prop is set.) That's meant for CRUD backends.

If you need fast, realtime events, you'd need to subscribe to the anno.store directly.

**Comment by rsimon** [comment]
About re-renders: this might get optimized eventually. But for now, every change to the annotation will trigger a re-render, which is by design. (The popup might want to react to changes  in the annotation.)

**Comment by technbuzz** [comment]
I think exposing certain api for each annotation so set the common svg styles i-e like fill, fillOpacity, stroke etc would. That's how Leaflet Circle Marker does. We could also allow user give a css class to have more control over the the svg path and so on.

But I found styling annotation as common use case.

**Comment by rsimon** [comment]
@technbuzz absolutely. Also see: https://annotorious.dev/api-reference/drawing-style/

P.S.: CSS classes directly are not an option for Annotorious, because the OpenSeadragon version works with WebGL rather than SVG.

**Comment by technbuzz** [comment]
By css classes i meant using this class to add svg properties, not using treating it to style html element. I am in full support of svg renderer in this scenario

**Comment by rsimon** [comment]
As I said: in Annotorious OpenSeadragon, there is no HTML element. It's WebGL-based.

**Comment by technbuzz** [comment]
Ah. okay. I misread it.

**Comment by technbuzz** [comment]
BTW I read your formatter tutorials which seems good way to cover custom styling and label. 
How we can add them to version 3?

**Comment by rsimon** [comment]
I would recommend sticking with Annotorious v2.7 for now, if that fits your needs. Porting this feature to WebGL is not trivial and will not happen anytime soon. Although it's more feasible for the standard image version of Annotorious (which still uses SVG), the standard image version isn't a priority for me right now. Which means I'm also not going to work on this anytime soon.

If you want to try fitting it into the standard image version yourself, the starting point would probably be the shape implementations: https://github.com/annotorious/annotorious/tree/main/packages/annotorious/src/annotation/shapes

**Comment by geo242** [comment]
Any update on this @rsimon? I have a client that is requesting persistent labels on the annotations and I don’t see a way to implement this in v3.

**Comment by rsimon** [comment]
Hi @geo242,

Sorry, no updates I'm afraid. It hasn't come up as a requirement for any of my clients so far. (I've actually closed [the relevant issue](https://github.com/annotorious/annotorious/issues/480) just recently due to lack of activity.) 

Depending on the exact requirements, I'd still recommend implementing this as a plugin rather than a core feature. But realistically, [as mentioned in the issue](https://github.com/annotorious/annotorious/issues/480#issuecomment-3432440656), that's very unlikely to happen until a sponsoring client comes along. If your client is open to funding the time, feel free to drop me a line at [hello@rainersimon.io](mailto:hello@rainersimon.io).

**Comment by rsimon** [comment]
Closing this since the conversation has drifted off from the original question. 
- Original question is still open as an issue here: #472 
- Persistent labels are not currently planned (see comment above)
